lib/io/QuadSinkHdt.ts

```diff
@@ -1,85 +1,123 @@
+    this.conversionConcurrency = options.conversionConcurrency ?? 1;
+    this.docker = new Docker();
+    this.filesToConvert = new Set();
   }
 
-  public async push(iri: string, quad: RDF.Quad): Promise<void> {
-    const path = Path.join('./', this.getFilePath(iri));
-    await super.push(iri, quad);
+  protected getFilePath(iri: string): string {
+    const path = super.getFilePath(iri);
+    this.filesToConvert.add(resolve(path));
+    return path;
+  }
+
+  protected async pullDockerImage(): Promise<void> {
+    await new Promise<void>((resolve, reject) => this.docker.pull(HDTCPP_DOCKER_IMAGE, {}, (error, result) => {
+      if (error) {
+        reject(error);
+      }
+      this.docker.modem.followProgress(result!, error => error ? reject(error) : resolve());
+    }));
+  }
 
-    // Add files with the defined extension to the list to be transformed
-    if (this.fileExtension !== undefined && path.includes(this.fileExtension)) {
-      this.files.add(path);
+  protected async convertSingleFile(rdfFilePath: string): Promise<void> {
+    const cmd = [
+      'rdf2hdt',
+      ...this.generateIndexes ? [ '-i' ] : [],
+      ...this.conversionProgress ? [ '-p' ] : [],
+      `${HDTCPP_MOUNT_PATH}/${basename(rdfFilePath)}`,
+      `${HDTCPP_MOUNT_PATH}/${basename(rdfFilePath).replace(this.fileExtension ?? '', '')}.hdt`,
+    ];
+    const createOptions: Docker.ContainerCreateOptions = {
+      // eslint-disable-next-line ts/naming-convention
+      HostConfig: {
+        // eslint-disable-next-line ts/naming-convention
+        AutoRemove: true,
+        // eslint-disable-next-line ts/naming-convention
+        Binds: [ `${dirname(rdfFilePath)}:${HDTCPP_MOUNT_PATH}:rw` ],
+      },
+    };
+    await this.docker.run(HDTCPP_DOCKER_IMAGE, cmd, process.stdout, createOptions);
+    if (this.removeSourceFiles) {
+      await unlink(rdfFilePath);
     }
   }
 
-  /**
-   * Log the number of files transformed into HDT
-   * @param {number} counter - counter of the number of files transformed
-   * @param {boolean} newLine - add a new line after the logging
-   */
-  private attemptLogHdtTransformation(counter: number, newLine = false): void {
-    if (this.log) {
-      readline.clearLine(process.stdout, 0);
-      readline.cursorTo(process.stdout, 0);
-      process.stdout.write(`\rfiles transformed to HDT:${counter} out of ${this.files.size}`);
-      if (newLine) {
-        process.stdout.write(`\n`);
+  protected async convertToHdt(): Promise<void> {
+    await this.pullDockerImage();
+    const conversionPromisePool = new Set<Promise<void>>();
+    let convertedFileCount = 0;
+    for (const rdfFilePath of this.filesToConvert) {
+      convertedFileCount++;
+      conversionPromisePool.add(this.convertSingleFile(rdfFilePath));
+      if (
+        conversionPromisePool.size === this.conversionConcurrency ||
+        convertedFileCount === this.filesToConvert.size
+      ) {
+        await Promise.all(conversionPromisePool);
```

Comment:
I think the promises should be raced so that the pool is always processing information instead of doing bash processing.

Reply:
This has been fixed now.